fix cpu autocast check in rnn

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100621


https://github.com/pytorch/pytorch/pull/100100 added Typechecking while `torch.is_autocast_enabled()` always return `False` on cpu. This PR fixes the autocast check for cpu.


cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

chunyuan-w committed:
fix cpu autocast check in rnn

[ghstack-poisoned]

chunyuan-w committed:
Update on "fix cpu autocast check in rnn"



https://github.com/pytorch/pytorch/pull/100100 added Typechecking while `torch.is_autocast_enabled()` always return `False` on cpu. This PR fixes the autocast check for cpu.


[ghstack-poisoned]

vadimkantorov commented:
Is ` torch.is_autocast_enabled() always return False on cpu.` by itself a bug though?

chunyuan-w commented:
> Is ` torch.is_autocast_enabled() always return False on cpu.` by itself a bug though?

@leslie-fang-intel could you please help with the above question?

leslie-fang-intel commented:
I think that's because different dispatch Key is used for CPU and CUDA, so we use different API to query the status.

vadimkantorov commented:
Well, I just hope that `torch.is_autocast_enabled()` does not always return False on CPU :) as bf16 and even fp16 autocast is getting increasingly supported on CPU...

shivamkc01 commented:
To fix the issue, the condition check in the RNN implementation needs to be updated to correctly handle this case. Specifically, the check should be modified to ensure that both the input and output tensors have the same device before performing the autocast check. This will ensure that the autocast check works correctly for all cases, including those where the input and output tensors have different devices.

_In my opinion below the line we can try._
--> `if not any(isinstance(h, torch.Tensor) and h.is_cuda for h in hidden) and input.is_cuda == output.is_cuda:`

Correct me if I'm wrong.

jgong5 commented:
> Is ` torch.is_autocast_enabled() always return False on cpu.` by itself a bug though?

That's a good question. Why `torch.is_autocast_enabled()` ignores CPU? From its API name, I would expect it return `True` intuitively when autocast is enabled for any device, not just CUDA.

chunyuan-w commented:
@leslie-fang-intel mentioned to me that there's an internal API `torch._C._is_any_autocast_enabled()` which seems to work for GPU, CPU and XPU:
https://github.com/pytorch/pytorch/blob/19a57870a3ef545fbf4c32ec2806dc89cdcdbea2/torch/csrc/autograd/init.cpp#L436-L445

It's still an internal API that is not exposed yet. @jgong5 is it better to use `torch._C._is_any_autocast_enabled()` in the condition check or we prefer to change the behavior of `torch.is_autocast_enabled()`?

chunyuan-w committed:
Update on "fix cpu autocast check in rnn"



https://github.com/pytorch/pytorch/pull/100100 added Typechecking while `torch.is_autocast_enabled()` always return `False` on cpu. This PR fixes the autocast check for cpu.


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

chunyuan-w commented:
> To fix the issue, the condition check in the RNN implementation needs to be updated to correctly handle this case. Specifically, the check should be modified to ensure that both the input and output tensors have the same device before performing the autocast check. This will ensure that the autocast check works correctly for all cases, including those where the input and output tensors have different devices.
> 
> _In my opinion below the line we can try._ --> `if not any(isinstance(h, torch.Tensor) and h.is_cuda for h in hidden) and input.is_cuda == output.is_cuda:`
> 
> Correct me if I'm wrong.

Thanks for the suggestion. This check seems a general requirement not specific to the RNN OP.
Currently, the change in https://github.com/pytorch/pytorch/pull/100100 has brought regression to the usage of lstm OP on CPU when CPU autocast is enabled, weight is in bf16 and input is in fp32.
If possible, I'd suggest implementing it in a separate PR if we meet cross-device issues in the future.

chunyuan-w commented:
@albanD could you please help review this PR?

albanD reviewed (commented):
Small change but this sounds good to me.
I feel like this API is long overdue for a cleanup (these functions are not even documented :( )
I'll open an issue about that later today.

chunyuan-w committed:
Update on "fix cpu autocast check in rnn"



https://github.com/pytorch/pytorch/pull/100100 added Typechecking while `torch.is_autocast_enabled()` always return `False` on cpu. This PR fixes the autocast check for cpu.


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

chunyuan-w commented on torch/nn/modules/rnn.py:
```diff
@@ -209,7 +209,7 @@ def reset_parameters(self) -> None:
 
     def check_input(self, input: Tensor, batch_sizes: Optional[Tensor]) -> None:
-        if input.dtype != self._flat_weights[0].dtype and not torch.is_autocast_enabled():
+        if input.dtype != self._flat_weights[0].dtype and (not torch.is_autocast_enabled() and not torch.is_autocast_cpu_enabled()):
```
Hi @albanD, I've tried with `torch._C._is_any_autocast_enabled()` but it seems that more changes are required if we want to use this API. The below error has been met in the CI:
```log
Python builtin <built-in function _is_any_autocast_enabled> is currently not supported in Torchscript
```

The previous change in https://github.com/pytorch/pytorch/pull/100100 has brought regression to the usage of lstm OP on CPU when CPU autocast is enabled, weight is in bf16 and input is in fp32.
May I know if it is possible to use the below way in this PR to fix the functionality on CPU and we can further refine the logic in follow-up PRs?
```python
not (torch.is_autocast_enabled() or torch.is_autocast_cpu_enabled())
```

albanD replied:
Ho interesting, did you try the following to avoid the jit issue? `and not torch.jit.is_scripting() and not torch._C._is_any_autocast_enabled()`

chunyuan-w replied:
I just tried for one of the failed cases locally `python -u test/test_jit.py -k test_nn_GRU`. The proposed code still raises error:
```
RuntimeError: 
Python builtin <built-in function _is_any_autocast_enabled> is currently not supported in Torchscript:
  File "/home/chunyuan/torch-inductor/pytorch/torch/nn/modules/rnn.py", line 217
            
        if input.dtype != self._flat_weights[0].dtype and not torch.jit.is_scripting() and not torch._C._is_any_autocast_enabled():
                                                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
```

The below one works:
```
if not torch.jit.is_scripting():
    if input.dtype != self._flat_weights[0].dtype and not torch._C._is_any_autocast_enabled():
        raise ValueError('input must have the type {}, got type {}'.format(
            self._flat_weights[0].dtype, input.dtype))
```

With this change, I'm wondering how we should handle the dtype check when `torch.jit.is_scripting() == True`?

chunyuan-w committed:
Update on "fix cpu autocast check in rnn"



https://github.com/pytorch/pytorch/pull/100100 added Typechecking while `torch.is_autocast_enabled()` always return `False` on cpu. This PR fixes the autocast check for cpu.


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

chunyuan-w committed:
Update on "fix cpu autocast check in rnn"



https://github.com/pytorch/pytorch/pull/100100 added Typechecking while `torch.is_autocast_enabled()` always return `False` on cpu. This PR fixes the autocast check for cpu.


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

chunyuan-w commented:
@pytorchbot merge

chunyuan-w commented:
@pytorchbot merge
